### PR TITLE
Improve naming of remove methods

### PR DESCRIPTION
**Pull request by mlodato517**
**This PR**
- Deprecates the `remove` method for a more descriptive method `remove_value`
- Deprecates the `empty` method for a more description method `remove_empty`
- Uses `swap_remove` in a test where the order doesn't matter so why not

**Why?**
Using the standard lib as a North Star, [`remove`](https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.remove) should be a method that takes a key and removes that key from the map. In `evmap`, `remove` is meant to take a key _and a value_ and remove _just that value_ from the value bag for the key. To make that more clear, we deprecate `remove` and rename it to `remove_value`.

Unfortunately we can't then rename `empty` (`evmap`'s version of the std `HashMap`s `remove`) to `remove` for backwards compatibility. Instead, we rename it to `remove_entry`.

---
This change is [Reviewable](https://reviewable.io/reviews/jonhoo/rust-evmap/59)

### CODE REVIEW

File: src/write.rs [hunk 1]
```diff
@@ -431,14 +431,30 @@ where
     /// Remove the given value from the value-bag of the given key.
     ///
     /// The updated value-bag will only be visible to readers after the next call to `refresh()`.
+    #[deprecated(since = "11.0.0", note = "Renamed to remove_value")]
```
Comment (mlodato517) [inline src/write.rs:434]: Is this the right version? I imagine the alpha you're on is therefore pre-`11.0.0`? In which case ... do we need to backwards compatible? I imagine it's a major release so ⚖️. Unless the current `11.0.0-alpha` is meant to be between `11.0.0` and `11.1.0` or some such?

File: src/write.rs [hunk 2]
```diff
@@ -431,14 +431,30 @@ where
...
     pub fn remove(&mut self, k: K, v: V) -> &mut Self {
+        self.remove_value(k, v)
```
Comment (mlodato517) [inline src/write.rs:436]: Is this the right way to deprecate this? So that if people are using the old one, they get new functionality? Not sure if that's good or not. Also not sure if this function should have its documentation removed so it doesn't get out of sync if `remove_value` changes? Or maybe I just shouldn't change the innards of `remove`?

File: src/write.rs [hunk 3]
```diff
@@ -431,14 +431,30 @@ where
...
+    }
+
+    /// Remove the given value from the value-bag of the given key.
+    ///
+    /// The updated value-bag will only be visible to readers after the next call to `refresh()`.
+    pub fn remove_value(&mut self, k: K, v: V) -> &mut Self {
         self.add_op(Operation::Remove(k, v))
```
Comment (mlodato517) [inline src/write.rs:443]: Should I also rename these operations? I imagine those are internal, implementation details and should be safe to rename? Although maybe not valuable ⚖️

File: tests/quick.rs [hunk 4]
```diff
@@ -110,16 +110,16 @@ fn do_ops<K, V, S>(
...
             }
             RemoveValue(ref k, ref v) => {
-                evmap.remove(k.clone(), v.clone());
+                evmap.remove_value(k.clone(), v.clone());
                 write_ref.get_mut(k).and_then(|values| {
                     values
                         .iter_mut()
                         .position(|value| value == v)
-                        .and_then(|pos| Some(values.remove(pos)))
+                        .and_then(|pos| Some(values.swap_remove(pos)))
```
Comment (mlodato517) [inline tests/quick.rs:122]: This is unrelated to this PR and is just cleanup from [this](https://github.com/jonhoo/rust-evmap/pull/58#issuecomment-654859029).

File: src/write.rs [hunk 5]
```diff
@@ -431,14 +431,30 @@ where
...
     /// The value-bag will only disappear from readers after the next call to `refresh()`.
+    #[deprecated(since = "11.0.0", note = "Renamed to remove_entry")]
     pub fn empty(&mut self, k: K) -> &mut Self {
+        self.remove_entry(k)
+    }
+
+    /// Remove the value-bag for the given key.
+    ///
+    /// The value-bag will only disappear from readers after the next call to `refresh()`.
+    pub fn remove_entry(&mut self, k: K) -> &mut Self {
```
Comment (mlodato517) [inline src/write.rs:457]: Oh, also, I later discovered that [`remove_entry`](https://doc.rust-lang.org/beta/std/collections/struct.HashMap.html#method.remove_entry) also exists for `std::collections::HashMap` so maybe we don't want to use this because this doesn't use the Entry API (yet?). Maybe this should be `remove_key` or `remove_bag` or something?

**Comment by mlodato517** [comment]
Oh, and that code coverage is a good point - I imagine we should still be testing deprecated methods?

File: src/write.rs (hunk 1 above)
Comment (jonhoo) [reply src/write.rs:434]: You're totally right that we don't _have_ to, since we're about to make a breaking release anyway. However, I think having one major version with the methods deprecated is worthwhile. Makes transitioning a little nicer :)

File: src/write.rs (hunk 2 above)
Comment (jonhoo) [reply src/write.rs:436]: There's no new functionality here though, right? Since it's just a rename, I think it's fine.

Comment (mlodato517) [reply src/write.rs:436]: Not yet! I was just thinking that if someone goes and updates `remove_value` for some reason, should those still using `remove` get that new change?

File: src/write.rs (hunk 5 above)
Comment (jonhoo) [reply src/write.rs:457]: It was precisely because of `HashMap::remove_entry` that I suggested the name :p The fact that it includes the key as well gives more of the flavor of "this removes the whole thing".

File: src/write.rs (hunk 3 above)
Comment (jonhoo) [reply src/write.rs:443]: Yes, I think renaming them is probably worthwhile.
I forget whether `Operation` is `pub` actually.. :/

**Comment by jonhoo** [comment]
Hmm... No, I think it's fine to leave them out.

File: src/write.rs (hunk 5 above)
Comment (mlodato517) [reply src/write.rs:457]: Haha okay! Yeah I don't think we need more words to describe "removes the whole thing" (as the expected API is probably just "remove") but I didn't know if we should reach for "not `_entry`" because one might expect `remove_entry` should return `Option<(K, V)>` and this doesn't.

File: src/write.rs (hunk 2 above)
Comment (jonhoo) [reply src/write.rs:436]: If `remove_value` gets updated without its semantics change, then yes, I think `remove` should also get that change. And changing the semantics of `remove_value` would be backwards-incompatbile anyway. So I think it's fine.

File: src/write.rs (hunk 5 above)
Comment (jonhoo) [reply src/write.rs:457]: Nah, I think it's probably fine the way it is. People already realize pretty quickly that `evmap` doesn't have the same interface as `HashMap`.

File: src/write.rs (hunk 3 above)
Comment (mlodato517) [reply src/write.rs:443]: Yeah it looks like `Operation` is `pub`. So I imagine we want to deprecate them instead of a straight rename? That means that we have to match on deprecated enum variants in a few places. Is that okay? Or do you want me to just delete those? And do I need to do any sort of "cargo ignore this match of a deprecated enum" so it's not super annoying?

Comment (jonhoo) [reply src/write.rs:443]: I think what I want to do is make `Operation` not be `pub`. That is a breaking change I think I'd be willing to make for `11.0`. If you change it to `pub(crate)`, how many things break?

Comment (mlodato517) [reply src/write.rs:443]: It looks like maybe just [`pending`](https://github.com/jonhoo/rust-evmap/blob/master/src/write.rs#L361)?

Comment (jonhoo) [reply src/write.rs:443]: Yeah, let's remove `pending` too. That's not really an API people should be using anyway.

Comment (mlodato517) [reply src/write.rs:443]: This is a dream come true. Love deleting code.

Comment (mlodato517) [reply src/write.rs:443]: Oh wait, you just meant make not public, I assume? Poor me 😢

Comment (mlodato517) [reply src/write.rs:443]: So for [this test](https://github.com/jonhoo/rust-evmap/blob/master/tests/lib.rs#L188) do I remove the assertion and just put a dummy `assert!(true)` after the `flush()` or do I move the test into `src/write`?

Comment (jonhoo) [reply src/write.rs:443]: I think you can actually just delete it. I don't think we use it internally for anything?
For that test, I'd move it to be a unit test, and then it can inspect the internal state that `pending()` accesses directly.

Comment (mlodato517) [reply src/write.rs]: So dfb4843184f9c8aa3e974e8607f753138ef5d53a makes `Operation` `pub(crate)`, removes `pending()`, and moves the test

File: src/write.rs [hunk 6]
```diff
@@ -336,38 +336,12 @@ where
...
-        &self.oplog[self.swap_index..]
-    }
-
     /// Refresh as necessary to ensure that all operations are visible to readers.
     ///
     /// `WriteHandle::refresh` will *always* wait for old readers to depart and swap the maps.
     /// This method will only do so if there are pending operations.
     pub fn flush(&mut self) -> &mut Self {
-        if !self.pending().is_empty() {
+        if !self.oplog[self.swap_index..].is_empty() {
```
Comment (mlodato517) [inline src/write.rs:344]: Now that we don't have `pending()` for the added readability, I kind of want to just check
```suggestion
        if self.swap_index < self.oplog.len() {
```
or whatever makes this work 🤔

Comment (jonhoo) [reply src/write.rs:344]: Yes, that should be right. Running the test suite should confirm it pretty quickly.

Comment (mlodato517) [reply src/write.rs:344]: I like the confidence! I wasn't sure because it looked like [only one test called `flush` and it didn't test much](https://github.com/jonhoo/rust-evmap/blob/master/tests/lib.rs#L189) but I'll go ahead with that change

Comment (mlodato517) [reply src/write.rs]: I guess I should maybe update/write some test to ensure that `refresh` isn't called if I call `flush` twice or something? Is that a thing that is possible to verify?

Comment (jonhoo) [reply src/write.rs]: Huh, interesting. I thought `flush` was called internally, but I guess not. Yes, then adding a test that checks that would be useful indeed! You'll probably need to write it as a unit test and then inspect some internal state to check if `refresh` was called.

Comment (mlodato517) [reply src/write.rs]: Added 351881d37f68326de12618ff8606969e76d1a876. I'm not sure if `w.second` is the best idea but there wasn't anything _terribly_ obvious

File: src/write.rs [hunk 7]
```diff
@@ -748,4 +748,46 @@ mod tests {
...
+    fn flush_no_refresh() {
+        let x = 'x';
+
+        let (_, mut w) = new();
+
+        // Until we refresh, writes are written directly instead of going to the
+        // oplog (because there can't be any readers on the w_handle table).
+        w.refresh();
+
+        assert_eq!(w.second, true);
```
Comment (jonhoo) [inline src/write.rs:762]: Maybe assert `!w.second` before the first `refresh`?

File: src/write.rs [hunk 8]
```diff
@@ -748,4 +748,46 @@ mod tests {
...
+
+        w.insert(x, x);
+        w.flush();
+
+        // A refresh happened!
+        assert_eq!(w.second, false);
+
+        // Reset this flag so we can see if another refresh happens
+        w.second = true;
```
Comment (jonhoo) [inline src/write.rs:776]: I wonder what kind of weird, deeper effects changing this externally might have... Hopefully none :sweat_smile:

Comment (mlodato517) [reply src/write.rs:776]: Haha yeah this feels ... horrible ... but I'm not sure how else to check, "Did a refresh happen"? Is there some other flag/meta data I could check?

Comment (jonhoo) [reply src/write.rs:776]: To be honest, I'd probably just add a:
```rust
#[cfg(test)]
refreshes: usize
```
field to `WriteHandle`, and then have a
```rust
#[cfg(test)]
self.refreshes += 1;
```
in the appropriate place in `refresh`. That way, in testing code, you'll know exactly how many refreshes have happened. And it wouldn't require hackery or impact running code.

Comment (mlodato517) [reply src/write.rs:776]: Hmmm maybe I just have the wrong syntax but I'm getting some
```
attributes on expressions are experimental
removing an expression is not supported in this position
```
when I try
```rust
pub struct WriteHandle {
    // ...
    #[cfg(test)]
    refreshes: usize,
}
pub(crate) fn new(/*...*/) -> WriteHandle {
    let m = w_handle.meta.clone();
    WriteHandle {
        // ...
        #[cfg(test)]
        refreshes: 0,
    }

  pub fn refresh(&mut self) -> &mut Self {
    // ...
    #[cfg(test)]
    self.refreshes += 1;
```
Was that what you had in mind?

Comment (jonhoo) [reply src/write.rs:776]: Ugh, I forgot they're still experimental. Just do this instead:
```rust
#[cfg(test)]
{
  self.refreshes += 1;
}
```
and that should work I think

Comment (mlodato517) [reply src/write.rs:776]: I'm a little wigged out about sprinkling the test cfg around too much but https://github.com/jonhoo/rust-evmap/pull/60 might do it
